server/src/game/role/marksman.rs

```diff
@@ -0,0 +1,208 @@
+
+use std::ops::Deref;
+
+use serde::Serialize;
+
+use crate::game::chat::{ChatGroup, ChatMessageVariant};
+use crate::game::grave::GraveKiller;
+use crate::game::phase::PhaseType;
+use crate::game::player::PlayerReference;
+use crate::game::role_list::Faction;
+use crate::game::visit::Visit;
+
+use crate::game::Game;
+use super::{Priority, RoleStateImpl, Role, RoleState};
+
+#[derive(Clone, Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Marksman {
+    state: MarksmanState
+}
+
+#[derive(Clone, Debug, Serialize, PartialEq, Eq)]
+#[serde(rename_all = "camelCase")]
+#[serde(tag = "type")]
+pub(self) enum MarksmanState{
+    NotLoaded,
+    Marks{
+        marks: MarksmanMarks
+    },
+    ShotTownie
+}
+impl MarksmanState{
+    fn no_marks(&self)->bool{
+        !matches!(
+            self,
+            MarksmanState::Marks{marks: MarksmanMarks::One{..}} |
+            MarksmanState::Marks{marks: MarksmanMarks::Two{..}}
+        )
+    }
+    fn marks(&self)->Box<[PlayerReference]> {
+        if let Self::Marks{marks} = self {
+            marks.marks()
+        }else{
+            Box::new([])   
+        }
+    }
+    /// This function will mark an unmarked player or un-mark a marked player
+    /// if the action is invalid, then it will do nothing
+    fn toggle_mark(&mut self, p: PlayerReference){
+        let Self::Marks { marks } = self else {return};
+        *marks = match marks {
+            MarksmanMarks::None => {
+                MarksmanMarks::One { a: p }
+            }
+            MarksmanMarks::One {a} => {
+                if *a != p {
+                    MarksmanMarks::Two { a: *a, b: p }
+                }else{
+                    MarksmanMarks::None
+                }
+            }
+            MarksmanMarks::Two { a, b } => {
+                if p == *a {
+                    MarksmanMarks::One { a: *b }
+                }else if p == *b {
+                    MarksmanMarks::One { a: *a }
+                }else{
+                    MarksmanMarks::Two { a: *a, b: *b }
+                }
+            }
+        };
+    }
+}
+#[derive(Clone, Debug, Serialize, PartialEq, Eq)]
+#[serde(rename_all = "camelCase")]
+#[serde(tag = "type")]
+enum MarksmanMarks{
+    None,
+    One{
+        a: PlayerReference
+    },
+    Two{
+        a: PlayerReference, 
+        b: PlayerReference
+    }
+}
+impl MarksmanMarks{
+    fn marks(&self)->Box<[PlayerReference]> {
+        match self {
+            MarksmanMarks::None => {Box::new([])},
+            MarksmanMarks::One{a} => {Box::new([*a])},
+            MarksmanMarks::Two{a,b} => {Box::new([*a,*b])}
+        }
+    }
+}
+
+impl Default for Marksman {
+    fn default() -> Self {
+        Self { state: MarksmanState::NotLoaded }
+    }
+}
```

Comment:
You can implement default for MarksmanState (look up rust enum default macro) and then auto-generate this using the derive macro